src/hotspot/share/prims/jvmtiThreadState.cpp

```diff
@@ -258,11 +281,45 @@ JvmtiVTMSTransitionDisabler::~JvmtiVTMSTransitionDisabler() {
   if (Thread::current_or_null() == NULL) {
     return;  // Detached thread, can be a call from Agent_OnLoad.
   }
-  enable_VTMS_transitions();
+  if (_vthread != NULL) {
+    VTMS_transition_enable_for_one(); // enable VTMS transitions for one virtual thread
+  } else {
+    VTMS_transition_enable_for_all(); // enable VTMS transitions for all virtual threads
+  }
+}
+
+// disable VTMS transitions for one virtual thread
+void
+JvmtiVTMSTransitionDisabler::VTMS_transition_disable_for_one() {
+  assert(_vthread != NULL, "sanity check");
+  JavaThread* thread = JavaThread::current();
+  HandleMark hm(thread);
+  Handle vth = Handle(thread, JNIHandles::resolve_external_guard(_vthread));
+  if (!java_lang_VirtualThread::is_instance(vth())) {
```

Comment:
How can this condition not be true? Should it be an assertion?

Reply:
The JvmtiVTMSTransitionDisabler has to be a no-op for a non-virtual thread.
It is to avoid different branches for virtual and non-virtual cases and use this in `jvmtiEnv.cpp`:
`  JvmtiVTMSTransitionDisabler disabler(thread);`

Reply:
I see. I assumed the name `_vthread` implied it was actually known to be a vthread. Really the field and parameters should just be called `thread`.

Reply:
Okay, renamed now.